modules/gdscript/gdscript_analyzer.cpp

```diff
@@ -1358,27 +1366,38 @@ void GDScriptAnalyzer::resolve_class_body(GDScriptParser::ClassNode *p_class, co
 
 					if (setter_function == nullptr) {
 						push_error(vformat(R"(Setter "%s" not found.)", member.variable->setter_pointer->name), member.variable);
-
-					} else if (setter_function->parameters.size() != 1) {
-						push_error(vformat(R"(Function "%s" cannot be used as setter because of its signature.)", setter_function->identifier->name), member.variable);
-
-					} else if (!is_type_compatible(member.variable->datatype, setter_function->parameters[0]->datatype, true)) {
-						push_error(vformat(R"(Function with argument type "%s" cannot be used as setter for a property of type "%s".)", setter_function->parameters[0]->datatype.to_string(), member.variable->datatype.to_string()), member.variable);
-
 					} else {
-						has_valid_setter = true;
-
+						if (setter_function->mandatory_parameter_count == 0 || setter_function->mandatory_parameter_count > 2) {
+							push_error(vformat(R"(Setters must have 1 or 2 mandatory arguments.)"), member.variable);
+						} else {
+							GDScriptParser::ParameterNode *value_arg = setter_function->parameters[setter_function->mandatory_parameter_count - 1];
+							if (!is_type_compatible(member.variable->datatype, value_arg->datatype, true)) {
+								push_error(vformat(R"(Value argument of type "%s" is not compatible with value type "%s".)", value_arg->datatype.to_string(), member.variable->datatype.to_string()), member.variable);
```

Comment:
```suggestion
								push_error(vformat(R"(Value argument of type "%s" is not compatible with variable type "%s".)", value_arg->datatype.to_string(), member.variable->datatype.to_string()), member.variable);
```

Also I really don't like the "value argument" even though I originally suggested it. Maybe we should just specify the argument name despite the long error message?

Reply:
That makes sense, considering it probably won't be clear what we mean by "value argument" and in a setter, it could be the first or second argument depending on whether it's a "named" setter or not.

I'll go with "Setter argument of type "%s" is not compatible with variable type "%s".